src/main/java/org/cloudfoundry/promregator/scanner/ReactiveAppInstanceScanner.java

```diff
@@ -192,23 +236,81 @@ public void setNumberOfInstances(int numberOfInstances) {
 
 			List<String> urls = sas.getUrls();
 			if (urls != null && !urls.isEmpty()) {
-				v.setAccessURL(this.determineAccessURL(v.getTarget().getProtocol(), urls, v.getTarget().getOriginalTarget().getPreferredRouteRegexPatterns(), v.getTarget().getPath()));
+				// Set the access url to the selected route (without any protocol or path yet)
+				v.setAccessURL(this.determineApplicationRoute(urls, v.getTarget().getOriginalTarget().getPreferredRouteRegexPatterns()));				
+      }      
+           
+			v.setNumberOfInstances(sas.getInstances());
+			
+			return Mono.just(v);
+    });
+        
+    Flux<List<RouteResource>> routeApplicationFlux = osaVectorSpaceFlux.flatMapSequential(v -> this.getAppRoutes(v.getApplicationId()));
+		Flux<OSAVector> osaVectorApplicationRouteFlux = Flux.zip(osaVectorApplicationFlux, routeApplicationFlux).flatMap(tuple -> {
+			OSAVector v = tuple.getT1();
+			
+			if (INVALID_ROUTES == tuple.getT2()) {
+				// NB: This drops the current target!
+				return Mono.empty();
 			}
 			
-			v.setNumberOfInstances(sas.getInstances());
+      List<RouteResource> routes = tuple.getT2();
+            
+      RouteResource metricsRoute = routes.stream()
+      .filter(r -> v.getAccessURL().contains(r.getUrl()))
+      .findFirst()
+			.orElse(null);
+
+			if (metricsRoute == null) {
+				return Mono.just(v);
+			}
+
+			Destination dest = metricsRoute.getDestinations()
+			.stream()
+			.filter(d -> d.getApplication().getApplicationId().equals(v.getApplicationId()))
+			.findFirst()
+			.orElse(null);
+						
+			if (dest != null) {
+				v.setInternalPort(dest.getPort());
+				v.setRouteId(metricsRoute.getId());
+				v.setDomainId(metricsRoute.getRelationships().getDomain().getData().getId());					
+			}
+            
+      return Mono.just(v);
+    });
+    
+		Flux<GetDomainResponse> domainFlux = osaVectorApplicationRouteFlux.flatMapSequential(v -> this.cfAccessor.retrieveDomain(v.getDomainId()));
```

Comment:
Performance impact: another roundtrip to the CFCC for each application. 
Moreover, the second roundtrip depends on the result of the first (added) roundtrip. 

Performance impact on this case might be significantly less, though, as our cache should have a rather high hit-ratio...

Reply:
Yes I think the domain trip should be very well cached. I can see how the application route round trip could add some overhead.

Reply:
I can again return a list of all domains for a particular space. This would reduce the chatter back and for for individual applications.